logic/transaction_pool.js

```diff
@@ -620,51 +627,54 @@ TransactionPool.prototype.queueTransaction = function(transaction, cb) {
 TransactionPool.prototype.undoUnconfirmedList = function(cb, tx) {
 	const ids = [];
 
-	async.eachSeries(
-		self.getUnconfirmedTransactionList(false),
-		(transaction, eachSeriesCb) => {
-			if (transaction) {
-				ids.push(transaction.id);
-				modules.transactions.undoUnconfirmed(
-					transaction,
-					err => {
-						// Remove transaction from unconfirmed, queued and multisignature lists
-						self.removeUnconfirmedTransaction(transaction.id);
-						if (err) {
-							library.logger.error(
-								`Failed to undo unconfirmed transaction: ${transaction.id}`,
-								err
-							);
-							return setImmediate(eachSeriesCb);
-						}
+	library.balancesSequence.add(
+		balancesSequenceCb => {
+			async.eachSeries(
+				self.getUnconfirmedTransactionList(false),
+				(transaction, eachSeriesCb) => {
+					if (transaction) {
```

Comment:
Same as above to filter the array first to avoid this condition and one nesting level.